### PR TITLE
[RESTEASY-1992] Prepare tests for extended support for multivalue @*Param

### PR DESCRIPTION
**Pull request by marekkopecky**
Add aditionall tests for extended support for multivalue @*Param

* Rename (and improve) original `MultiValuedParamTest` to `MultiValuedParamDateProxyTest`
   * This test uses java.util.Date class as multivalue param
   * This test uses proxy (String parameter is defined in proxy, server uses collections in end-point), proxy is not related with this feature
* `MultiValuedParamDateNoProxyTest`
     * Same as `MultiValuedParamDateProxyTest`, but proxy is not used at all in this test
* `MultiValuedParamCustomClassTest`
    * This test uses custom deployment class
* `MultiValuedParamCdiTest`
    * Server end-point is defined as a CDI bean with @RequestScoped annotation

Jira: https://issues.jboss.org/browse/RESTEASY-1992
3.6.1 PR: https://github.com/resteasy/Resteasy/pull/1660
Upstream PR: https://github.com/resteasy/Resteasy/pull/1659

### CODE REVIEW

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCdiTest.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.runner.RunWith;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.core.Response;
+
+/**
+ * @tpSubChapter Parameters
+ * @tpChapter Integration tests
+ * @tpTestCaseDetails Test of CDI integration for RESTEasy extended support for multivalue @*Param (RESTEASY-1566 + RESTEASY-1746)
+ * @tpSince RESTEasy 3.6.1.Final
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCdiTest.java]: 3.6.2.Final, I think.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCdiTest.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    static {
+        person1.setName(name1);
+        person2.setName(name2);
+        person3.setName(name3);
+        expectedResponse = person1 + "," + person2 + "," + person3;
+    }
+
+    /**
+     * @tpTestDetails Check queryParam in CDI integration
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCdiTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import javax.ws.rs.core.Response;
+
+/**
+ * @tpSubChapter Parameters
+ * @tpChapter Integration tests
+ * @tpTestCaseDetails Test for RESTEasy extended support for multivalue @*Param (RESTEASY-1566 + RESTEASY-1746)
+ *                    org.jboss.resteasy.test.resource.param.resource.MultiValuedParamPersonWithConverter class is used
+ * @tpSince RESTEasy 3.6.1.Final
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: 3.6.2.Final

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    private String generateBaseUrl() {
+        return PortProviderUtil.generateBaseUrl(MultiValuedParamCustomClassTest.class.getSimpleName());
+    }
+
+
+    /**
+     * Define testcase data set
+     */
+    static String name1 = "George";
+    static String name2 = "Jack";
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: Should be "Ringo". ;-)

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        person1.setName(name1);
+        person2.setName(name2);
+        person3.setName(name3);
+        expectedResponse = person1 + "," + person2 + "," + person3;
+    }
+
+
+    /**
+     * @tpTestDetails QueryParam test
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+
+        } finally {
+            client.close();
+        }
+    }
+
+
+    /**
+     * @tpTestDetails HeaderParam test
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+            response.close();
+
+        } finally {
+            client.close();
+        }
+    }
+
+    /**
+     * @tpTestDetails MatrixParam test
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+            response.close();
+
+        } finally {
+            client.close();
+        }
+    }
+
+    /**
+     * @tpTestDetails CookieParam test
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+            response.close();
+
+        } finally {
+            client.close();
+        }
+    }
+
+    /**
+     * @tpTestDetails PathParam test
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+
+/**
+ * @tpSubChapter Parameters
+ * @tpChapter Integration tests
+ * @tpTestCaseDetails Test for RESTEasy extended support for multivalue @*Param (RESTEASY-1566 + RESTEASY-1746)
+ *                    java.util.Date class is used
+ *                    Client Proxy is not used
+ * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+
+    public MultiValuedParamDateNoProxyTest() {
+        dates.add("date", date1);
+        dates.add("date", date2);
+        dates.add("date", date3);
+    }
+
+    /**
+     * @tpTestDetails Set specific values
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+
+        } finally {
+            client.close();
+        }
+    }
+
+
+    /**
+     * @tpTestDetails Set specific values
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+            response.close();
+
+        } finally {
+            client.close();
+        }
+    }
+
+    /**
+     * @tpTestDetails Set specific values
+     * @tpSince RESTEasy 3.6.1
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamDateNoProxyTest.java]: 3.6.2

**Comment by ronsigal** [comment]
@marekkopecky, aren't all (well, nearly all) classes managed beans now?

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+     */
+    @Test
+    public void testCookieParam() throws Exception {
+        ResteasyClient client = new ResteasyClientBuilder().build();
+        try {
+            Response response;
+            String cookieExpectedResponse = person1.toString();
+
+            response = client.target(generateBaseUrl() + "/cookieParam/customConversion_list")
+                    .request().cookie("person",name1).get();
```
Comment (ronsigal) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: Why just one person?

**Comment by ronsigal** [comment]
@marekkopecky , very nice.

**Comment by marekkopecky** [comment]
`Aren't all (well, nearly all) classes managed beans now?`

Well, I'm not an CDI expert, but AFAIK, just classes with "Bean defining annotations" are managed beans (see [the CDI spec, chapter 2.5.1](docs.jboss.org/cdi/spec/2.0/cdi-spec.pdf)) or beans can be defined in beans.xml. Anyway, the point of the `MultiValuedParamCdiTest` test is that user set(/change) scope of end-point. `MultiValuedParamCdiTest` uses `RequestScoped` annotation. Default CDI scope is Dependent ...

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java (hunk 17 above)
Comment (marekkopecky) [reply testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/param/MultiValuedParamCustomClassTest.java]: My mistake, of course more persons can be used here ...

**Comment by marekkopecky** [comment]
@ronsigal PR updated, thank you for the review!

**Pull request merged**